BaseClasses.py

```diff
@@ -759,20 +742,117 @@ def sweep_for_events(self, locations: Optional[Iterable[Location]] = None) -> No
+        """
+        Sweep through the locations that contain uncollected advancement items, collecting the items into the state
+        until there are no more reachable locations that contain uncollected advancement items.
+
+        :param locations: The locations to sweep through, defaulting to all locations in the multiworld.
+        :param yield_each_sweep: When True, return a generator that yields at the end of each sweep iteration.
+        :param checked_locations: Optional override of locations to filter out from the locations argument, defaults to
+        self.advancements when None.
+        """
+        if checked_locations is None:
+            checked_locations = self.advancements
+
+        # Since the sweep loop usually performs many iterations, the locations are filtered in advance.
+        # A list of tuples is used, instead of a dictionary, because it is faster to iterate.
+        advancements_per_player: List[Tuple[int, List[Location]]]
         if locations is None:
-            locations = self.multiworld.get_filled_locations()
-        reachable_advancements = True
-        # since the loop has a good chance to run more than once, only filter the advancements once
-        locations = {location for location in locations if location.advancement and location not in self.advancements}
-
-        while reachable_advancements:
-            reachable_advancements = {location for location in locations if location.can_reach(self)}
-            locations -= reachable_advancements
-            for advancement in reachable_advancements:
-                self.advancements.add(advancement)
-                assert isinstance(advancement.item, Item), "tried to collect Event with no Item"
-                self.collect(advancement.item, True, advancement)
+            # `location.advancement` can only be True for filled locations, so unfilled locations are filtered out.
+            advancements_per_player = []
+            for player, locations_dict in self.multiworld.regions.location_cache.items():
+                filtered_locations = [location for location in locations_dict.values()
+                                      if location.advancement and location not in checked_locations]
+                if filtered_locations:
+                    advancements_per_player.append((player, filtered_locations))
+        else:
+            # Filter and separate the locations into a list for each player.
+            advancements_per_player_dict: Dict[int, List[Location]] = defaultdict(list)
+            for location in locations:
+                if location.advancement and location not in checked_locations:
+                    advancements_per_player_dict[location.player].append(location)
+            # Convert to a list of tuples.
+            advancements_per_player = list(advancements_per_player_dict.items())
+            del advancements_per_player_dict
```

Comment:
How often do you expect this to be only a small subset of all players? I would just directly construct one tuple for each player anyway.

```suggestion
            advancements_per_player: List[Tuple[int, List[Location]]] = [
                (player_id, []) for player_id in self.multiworld.player_ids
            ]
            for location in locations:
                if location.advancement and location not in checked_locations:
                    advancements_per_player[location.player - 1][1].append(location)
```

Reply:
I expect it to be most of the time because the vast majority of sweeping does not use this conditional branch to begin with. Progression fill (`Fill.distribute_items_restrictive()` calling `Fill.fill_restrictive()`) and playthrough calculation (`Spoiler.create_playthrough()` calling `MultiWorld.can_beat_game()`), both pass in `locations=None` and thus do not run this conditional branch.

A world's `pre_fill()` often sweeps with only the locations belonging to the current player (`Fill.fill_restrictive(single_player_placement=True)`).
`Fill.distribute_early_items` also does a sweep limited to event (`location.address is None`) locations and only some games use events.
It looks like progression balancing also does some sweeping limited to locations belonging to a specific player, but I haven't looked into progression balancing in much detail yet.